jetty-server/src/main/java/org/eclipse/jetty/server/session/Session.java

```diff
@@ -1043,7 +1044,7 @@ protected void finishInvalidate() throws IllegalStateException
                     do
                     {
                         keys = _sessionData.getKeys();
-                        for (String key : keys)
+                        for (String key : new HashSet<>(keys))
```

Comment:
Thanks for the PR.  I think the better way to solve this would be to ensure that SessionData.getKeys() always returns an unmodifiable set. Can you update your PR?